test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/SidecarWatcher.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (C) 2022 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.services.bdd.suites.contract.traceability;
+
+import com.hedera.services.recordstreaming.RecordStreamingUtils;
+import com.hedera.services.stream.proto.SidecarFile;
+import com.hedera.services.stream.proto.TransactionSidecarRecord;
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.regex.Pattern;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class SidecarWatcher {
+    private static final Logger log = LogManager.getLogger(SidecarWatcher.class);
+    private static final Pattern SIDECAR_FILE_REGEX =
+            Pattern.compile("\\d{4}-\\d{2}-\\d{2}T\\d{2}_\\d{2}_\\d{2}\\.\\d{9}Z_\\d{2}.rcd");
+
+    private final Queue<Pair<String, TransactionSidecarRecord>> expectedSidecars =
+            new LinkedBlockingDeque<>();
+    private final Map<String, List<Pair<TransactionSidecarRecord, TransactionSidecarRecord>>>
+            failedSidecars = new HashMap<>();
+
+    private boolean shouldTerminateAfterNext = false;
+    private boolean hasSeenFirst = false;
+
+    public void startWatching(String s) throws IOException {
+        final var watchService = FileSystems.getDefault().newWatchService();
+        final var path = Paths.get(s);
+        path.register(watchService, StandardWatchEventKinds.ENTRY_CREATE);
```

Comment:
Looks like enabling both `ENTRY_CREATE` and `ENTRY_MODIFY` makes us receive 2 events for the same file (one time for the creation and one time for the modification of the folder, as you pointed out). IMO we should leave only one, not sure if it will make a difference in our case which one we choose?